staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go

```diff
@@ -117,25 +117,63 @@ type DiffOptions struct {
 	pruner           *pruner
 }
 
-func NewDiffOptions(ioStreams genericclioptions.IOStreams) *DiffOptions {
-	return &DiffOptions{
+// DiffFlags directly reflect the information that CLI is gathering via flags.  They will be converted to Options, which
+// reflect the runtime requirements for the command.  This structure reduces the transformation to wiring and makes
+// the logic itself easy to unit test
+type DiffFlags struct {
+	FilenameOptions resource.FilenameOptions
+
+	ServerSideApply   bool
+	FieldManager      string
+	ForceConflicts    bool
+	ShowManagedFields bool
+
+	Selector string
+	Diff     *DiffProgram
+}
+
+// NewDiffFlags returns a default DiffFlags
+func NewDiffFlags(ioStreams genericclioptions.IOStreams) *DiffFlags {
+	return &DiffFlags{
 		Diff: &DiffProgram{
 			Exec:      exec.New(),
 			IOStreams: ioStreams,
 		},
 	}
 }
 
+// AddFlags registers flags for a cli
+func (f *DiffFlags) AddFlags(cmd *cobra.Command) {
+	// Flag errors exit with code 1, however according to the diff
+	// command it means changes were found.
+	// Thus, it should return status code greater than 1.
+	cmd.SetFlagErrorFunc(func(command *cobra.Command, err error) error {
```

Comment:
It would be better that `AddFlags` only contains adding flags, this function should return to it's own place.

Reply:
I agree.